packages/palette/src/utils/useFocusLock.ts

```diff
@@ -0,0 +1,88 @@
+import React, { useCallback, useEffect, useState } from "react"
+import { useCursor } from "use-cursor"
+import { useMutationObserver } from "./useMutationObserver"
+
+const FOCUSABLE_SELECTOR = [
+  "a[href]",
+  "area[href]",
+  "input:not([disabled])",
+  "select:not([disabled])",
+  "textarea:not([disabled])",
+  "button:not([disabled])",
+  '[tabindex="0"]',
+].join(", ")
+
+/**
+ * Locks focus within the given element
+ */
+export const useFocusLock = (
+  ref: React.MutableRefObject<HTMLElement | null>
+) => {
+  const [focusableEls, setFocusableEls] = useState<HTMLElement[]>([])
+
+  const updateFocusableEls = useCallback(() => {
+    if (ref.current === null) return
+
+    setFocusableEls(
+      Array.from(ref.current.querySelectorAll<HTMLElement>(FOCUSABLE_SELECTOR))
+    )
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  useEffect(updateFocusableEls, [updateFocusableEls])
+
+  useMutationObserver({
+    ref,
+    onMutate: (mutations) => {
+      // Check to see if any of the mutations has either added or removed nodes
+      const hasMeaningfullyMutated = mutations.some((mutation) => {
+        return (
+          mutation.addedNodes.length > 0 || mutation.removedNodes.length > 0
+        )
+      })
+
+      if (hasMeaningfullyMutated) {
+        updateFocusableEls()
+      }
+    },
+  })
+
+  const { index: focusableIndex, handlePrev, handleNext } = useCursor({
+    max: focusableEls.length,
+  })
+
+  useEffect(() => {
+    if (!focusableEls.length) return
+
+    // Moves focus when index changes
+    focusableEls[focusableIndex].focus()
+  }, [focusableEls, focusableIndex])
+
+  useEffect(() => {
+    const handleKeydown = (event: KeyboardEvent) => {
+      switch (event.key) {
+        case "Tab":
+          // Lock focus within modal
+          event.preventDefault()
+          event.stopPropagation()
+
+          // Move focus up or down
+          event.shiftKey ? handlePrev() : handleNext()
+          break
+        default:
+          break
+      }
+    }
+
+    const focusedElPriorToOpen = document.activeElement as HTMLElement
+
+    document.addEventListener("keydown", handleKeydown)
+
+    return () => {
+      // Return the focus
+      focusedElPriorToOpen.focus()
+
+      document.removeEventListener("keydown", handleKeydown)
+    }
+  }, [handleNext, handlePrev])
+}
```

Comment:
😅 - I know you explained it a bit in standup, but whats the usecase for this again? I think I follow but curious if you could elaborate a bit.

Reply:
Just a basic usability/accessibility thing with modals; you want to trap focus within them. Consider what'd happen if you tabbed past the last element in a modal when it wasn't locked within (which is what happens with our current login/sign up modal btw): focus moves into a part of the page that you can't click without closing the modal.

This hook locks focus within the given element, so we pass it the modal container and then while the modal is open you can only tab around to things within the dialog. When the dialog is closed, focus is yielded back to the last thing that was focused on the page.